core/src/main/java/org/jboss/pnc/rex/core/GenericVertxHttpClient.java

```diff
@@ -148,24 +154,37 @@ private Uni<HttpResponse<Buffer>> handleRequest(Uni<HttpResponse<Buffer>> uni, C
                 );
 
         // cases when http request method itself fails (unreachable host)
-        uni = uni.onFailure(this::skipOn425)
+        uniWithCtx = uniWithCtx.onFailure(this::skipOnResponse)
                 .invoke(t ->  {
                     if (t instanceof RequestRetryException) {
                         log.warn("HTTP-CLIENT : Http call failed. RETRYING.");
-                    } else {
-                        log.warn("HTTP-CLIENT : Http call failed. RETRYING. Reason: ", t);
+                    } else if (t instanceof HttpResponseException hre) {
+                        log.warn("HTTP-CLIENT : Http call failed ({}). RETRYING.", hre.getStatusCode());
                     }
                 });
-
-        uni = requestRetryPolicy
-                .applyToleranceOn(this::skipOn425, uni);
+        uniWithCtx = requestRetryPolicy.applyToleranceOn(this::skipOnResponse, uniWithCtx);
 
         // fallback to connection unreachable after an exception
-        uni = uni.onFailure().invoke(onConnectionUnreachable)
+        BiFunction<Throwable, Context, Void> onError = (t, ctx) -> {
+            if (!(t instanceof HttpResponseException) && t.getCause() instanceof HttpResponseException) {
+                // prevent java.lang.IllegalStateException in case of exhausted retries due to expire-in limit
+                onConnectionUnreachable.accept(t.getCause());
+            } else {
+                onConnectionUnreachable.accept(t);
+            }
+            HttpResponse<Buffer> response = ctx.get(CTX_RESPONSE_KEY);
```

Comment:
I don't get this. This fallback is used when the http client itself throws an exception (can't reach, network error). I shouldn't have any Http response. I think this fallback is  skipped on Response Errors because this line is just before it:
`uniWithCtx = requestRetryPolicy.applyToleranceOn(this::skipOnResponse, uniWithCtx)`
So it shouldn't get to this point

Reply:
It is skipped on response errors, but just until the retries are not exhausted.

Reply:
Right, makes sense!